src/Avalonia.Controls/DockPanel.cs

```diff
@@ -75,132 +92,186 @@ public bool LastChildFill
+                        if (childConstraint.Width is not 0)
+                            accumulatedWidth += HorizontalSpacing;
+                        accumulatedWidth += childDesiredSize.Width;
+                        break;
+
                     case Dock.Right:
+                        rightSpacing = true;
                         parentHeight = Math.Max(parentHeight, accumulatedHeight + childDesiredSize.Height);
+                        if (childConstraint.Width is not 0)
+                            accumulatedWidth += HorizontalSpacing;
                         accumulatedWidth += childDesiredSize.Width;
                         break;
 
                     case Dock.Top:
+                        topSpacing = true;
+                        parentWidth = Math.Max(parentWidth, accumulatedWidth + childDesiredSize.Width);
+                        if (childConstraint.Height is not 0)
+                            accumulatedHeight += VerticalSpacing;
+                        accumulatedHeight += childDesiredSize.Height;
+                        break;
+
                     case Dock.Bottom:
+                        bottomSpacing = true;
                         parentWidth = Math.Max(parentWidth, accumulatedWidth + childDesiredSize.Width);
+                        if (childConstraint.Height is not 0)
+                            accumulatedHeight += VerticalSpacing;
                         accumulatedHeight += childDesiredSize.Height;
                         break;
                 }
             }
 
-            // Make sure the final accumulated size is reflected in parentSize.
-            parentWidth = Math.Max(parentWidth, accumulatedWidth);
-            parentHeight = Math.Max(parentHeight, accumulatedHeight);
+            if (LastChildFill)
+            {
+                var child = Children[Children.Count - 1];
+                var childConstraint = new Size(
+                    GetPositiveOrZero(availableSize.Width - accumulatedWidth),
+                    GetPositiveOrZero(availableSize.Height - accumulatedHeight));
+
+                child.Measure(childConstraint);
+                var childDesiredSize = child.DesiredSize;
+                parentHeight = Math.Max(parentHeight, accumulatedHeight + childDesiredSize.Height);
+                parentWidth = Math.Max(parentWidth, accumulatedWidth + childDesiredSize.Width);
+                accumulatedHeight += childDesiredSize.Height;
+                accumulatedWidth += childDesiredSize.Width;
+            }
+            else
+            {
+                if (leftSpacing || rightSpacing)
+                    accumulatedWidth -= HorizontalSpacing;
+                if (bottomSpacing || topSpacing)
+                    accumulatedHeight -= VerticalSpacing;
+            }
 
-            return (new Size(parentWidth, parentHeight));
+            parentWidth = Math.Min(availableSize.Width, Math.Max(parentWidth, accumulatedWidth));
```

Comment:
The clamp to `availableSize` wasn't present before, please remove it. This will prevent a parent from knowing what's the real desired size for the control.